File: apps/fsharp/src/MachineLearning/Chapter14/MlNetKMeans.fs
module MachineLearning.Chapter14.MlNetKMeans

open Microsoft.ML
open Microsoft.ML.Data
open Microsoft.ML.Trainers
open MachineLearning.Chapter14.KMeans

/// ML.NET に渡す 1 行
[<CLIMutable>]
type PointRow = { Features: float32[] }

/// ML.NET の予測。クラスタ番号は 1 から始まる
[<CLIMutable>]
type ClusterPrediction = { PredictedLabel: uint32 }

/// ML.NET の K-means（初期中心は k-means++ で ML.NET が選ぶ）でクラスタリングし、クラスタ番号（0 から始まる）を返す
let private clusterWithMlNet (nClusters: int) (seed: int) (points: Point[]) : int[] =
    let context = MLContext(seed = seed)
    let schema = SchemaDefinition.Create(typeof<PointRow>)
    schema["Features"].ColumnType <- VectorDataViewType(NumberDataViewType.Single, points[0].Length)

    let data =
        context.Data.LoadFromEnumerable(points |> Array.map (fun p -> { Features = Array.map float32 p }), schema)

    let options =
        KMeansTrainer.Options(
            NumberOfClusters = nClusters,
            InitializationAlgorithm = KMeansTrainer.InitializationAlgorithm.KMeansPlusPlus,
            NumberOfThreads = System.Nullable 1
        )

    let model = context.Clustering.Trainers.KMeans(options).Fit(data)

    context.Data.CreateEnumerable<ClusterPrediction>(model.Transform data, reuseRowObject = false)
    |> Seq.map (fun prediction -> int prediction.PredictedLabel - 1)
    |> Seq.toArray

/// シードを 1 ずつずらして nInit 回クラスタリングし、自作と同じ式で求めた SSE の最小値を返す。
/// 中心は、ML.NET が割り当てたクラスタごとの点の平均とする
let mlNetBestSse (nClusters: int) (seed: int) (nInit: int) (points: Point[]) : float =
    [ seed .. seed + nInit - 1 ]
    |> List.map (fun s ->
        let labels = clusterWithMlNet nClusters s points
        let centers = updateCenters points labels (Array.replicate nClusters points[0])
        sumOfSquaredErrors points labels centers)
    |> List.min

File: apps/csharp/src/MachineLearning/Chapter14/MlNetKMeans.cs
namespace MachineLearning.Chapter14;

using Microsoft.ML;
using Microsoft.ML.Data;
using Microsoft.ML.Trainers;
using Point = System.Collections.Generic.IReadOnlyList<double>;

/// <summary>ML.NET に渡す 1 行。列の長さは実行時に決まるので、SchemaDefinition で型を指定する。</summary>
public sealed class PointRow
{
    public float[] Features { get; set; } = [];
}

/// <summary>ML.NET の予測。クラスタ番号は 1 から始まる。</summary>
public sealed class ClusterPrediction
{
    public uint PredictedLabel { get; set; }
}

/// <summary>
/// ML.NET の K-means でクラスタリングする。初期中心は渡せないので（<see cref="KMeansTrainer.Options"/> には
/// 初期化のアルゴリズムを選ぶ設定しか無い）、自作と同じ初期中心での突き合わせはできない。SSE だけを比べる。
/// </summary>
public static class MlNetKMeans
{
    /// <summary>クラスタリングして、0 から始まるクラスタ番号を返す。</summary>
    public static IReadOnlyList<int> Cluster(int nClusters, int seed, IReadOnlyList<Point> points)
    {
        ArgumentNullException.ThrowIfNull(points);
        var context = new MLContext(seed: seed);
        var schema = SchemaDefinition.Create(typeof(PointRow));
        schema["Features"].ColumnType = new VectorDataViewType(NumberDataViewType.Single, points[0].Count);
        var data = context.Data.LoadFromEnumerable(
            points.Select(point => new PointRow { Features = [.. point.Select(value => (float)value)] }).ToList(),
            schema);
        var options = new KMeansTrainer.Options
        {
            NumberOfClusters = nClusters,
            InitializationAlgorithm = KMeansTrainer.InitializationAlgorithm.KMeansPlusPlus,
            NumberOfThreads = 1,
        };
        var model = context.Clustering.Trainers.KMeans(options).Fit(data);
        return [.. context.Data
            .CreateEnumerable<ClusterPrediction>(model.Transform(data), reuseRowObject: false)
            .Select(prediction => (int)prediction.PredictedLabel - 1)];
    }

    /// <summary>
    /// シードを 1 ずつずらして nInit 回クラスタリングし、自作と同じ式で求めた SSE の最小値を返す。
    /// 中心は、ML.NET が割り当てたクラスタごとの点の平均とする。
    /// </summary>
    public static double BestSse(int nClusters, int seed, int nInit, IReadOnlyList<Point> points)
    {
        ArgumentNullException.ThrowIfNull(points);
        return Enumerable.Range(seed, nInit).Min(s =>
        {
            var labels = Cluster(nClusters, s, points);
            var centers = KMeans.UpdateCenters(points, labels, [.. Enumerable.Repeat(points[0], nClusters)]);
            return KMeans.SumOfSquaredErrors(points, labels, centers);
        });
    }
}
